src/sonic-eventd/src/eventd.cpp

```diff
@@ -0,0 +1,499 @@
+     * is empty, do this check.
+     */
+    init_cnt = (int)m_events.size();
+
+    /* Read until STOP_CAPTURE */
+    while(m_ctrl == START_CAPTURE) {
+        runtime_id_t rid;
+        sequence_t seq;
+        internal_event_t event;
+        string source, evt_str;
+
+        if ((rc = zmq_message_read(sock, 0, source, event)) != 0) {
+            /*
+             * The capture socket captures SUBSCRIBE requests too.
+             * The messge could contain subscribe filter strings and binary code.
+             * Empty string with binary code will fail to deserialize.
+             * Else would fail event validation.
+             */
+            RET_ON_ERR((rc == EAGAIN) || (rc == ERR_MESSAGE_INVALID),
+                "0:Failed to read from capture socket");
+            continue;
+        }
+        if (!validate_event(event, rid, seq)) {
+            continue;
+        }
+        serialize(event, evt_str);
+
+        switch(cap_state) {
+        case CAP_STATE_INIT:
+            {
+                bool add = true;
+                init_cnt--;
+                pre_exist_id_t::iterator it = m_pre_exist_id.find(rid);
+
+                if (it != m_pre_exist_id.end()) {
+                    if (seq <= it->second) {
+                        /* Duplicate; Later/same seq in cache. */
+                        add = false;
+                    }
+                    if (seq >= it->second) {
+                        /* new one; This runtime ID need not be checked again */
+                        m_pre_exist_id.erase(it);
+                    }
+                }
+                if (add) {
+                    m_events.push_back(evt_str);
+                }
+            }
+            if(m_pre_exist_id.empty() || (init_cnt <= 0)) {
+                /* Init check is no more needed. */
+                pre_exist_id_t().swap(m_pre_exist_id);
+                cap_state = CAP_STATE_ACTIVE;
+            }
+            break;
+
+        case CAP_STATE_ACTIVE:
+            /* Save until max allowed */
+            try
+            {
+                m_events.push_back(evt_str);
+                if (VEC_SIZE(m_events) >= m_cache_max) {
+                    cap_state = CAP_STATE_LAST;
+                }
+                break;
+            }
+            catch (bad_alloc& e) 
+            {
+                stringstream ss;
+                ss << e.what();
+                SWSS_LOG_ERROR("Cache save event failed with %s events:size=%d",
+                        ss.str().c_str(), VEC_SIZE(m_events));
+                cap_state = CAP_STATE_LAST;
+                // fall through to save this event in last set.
+            }
+
+        case CAP_STATE_LAST:
+            if (!m_last_events_init) {
+                /* Clear the map, created to ensure memory space available */
+                m_last_events.clear();
+                m_last_events_init = true;
+            }
+            m_last_events[rid] = evt_str;
```

Comment:
is rid value always from 0 to 999?

Reply:
It is string. uuid_t

Reply:
No. We save unlimited. 
STL::map does not allow reserve. So I try to pre-allocate dummy map of some max entries, just to have some mem available, if in  case our MAX-CACHE-SIZE exhausts memory. If you notice, we clear it at the start. This clear will return this free memory, either to OS or could be kept by STL code. In either case, when we start to add entries to m_last_Events, there is likely free mem to use. Just a work around for lack of mem reserve.

Reply:
From my earlier comment, we are putting the max entries of 1000 into the m_last_event just to reserve and clearing it here so we have at least that much space available when we want to write to it. That is the understanding?

Reply:
Yes.